src/mscorlib/src/System/Collections/Generic/List.cs

```diff
@@ -1031,6 +1028,45 @@ public void Sort(IComparer<T> comparer)
             return true;
         }
 
+        private void AddEnumerable(IEnumerable<T> enumerable)
+        {
+            Contract.Assert(enumerable != null);
+            Contract.Assert(!(enumerable is ICollection<T>), "We should have optimized for this beforehand.");
+
+            using (IEnumerator<T> en = enumerable.GetEnumerator())
+            {
+                T[] items = _items;
+                int size = _size;
+
+                // Read in items from the enumerator, only updating fields
+                // when we run out of space.
+
+                try
+                {
+                    while (en.MoveNext())
+                    {
+                        if (size == items.Length)
+                        {
+                            _size = size;
+                            EnsureCapacity(size + 1);
+                            items = _items;
+                        }
+
+                        // Note: It's important we increment size after Current is called.
+                        // If that throws an exception we don't want to do the increment.
+                        items[size] = en.Current;
+                        size++;
+                    }
+                }
+                finally
+                {
+                    // Make a final update to _size and _version after we've finished adding.
+                    _size = size;
+                    _version++;
```

Comment:
> AFAICS caching _items is harmless though

How so?  If the iterator adds an element that causes the list to grow, ```_items``` will end up pointing to a different array than ```items``` points to, and AddRange will start storing the values into a garbage array.